File: 백준/Gold/13910. 개업/개업.py
"""
관찰
양 손으로 만들 수 있다
한 손으로도 만들 수 있다

그럼 양손으로 만들 수 있는 모든 경우의 수를 웍 리스트에 넣기

만들 수 있는 모든 웍 리스트로 bfs 하기
"""
import sys
from collections import deque

input = lambda: sys.stdin.readline().strip()

N, M = map(int, input().split())
wok = list(map(int, input().split()))

all_wok = wok[:]
for i in range(M):
    for j in range(i, M):
        if i == j:
            continue
        if wok[i] + wok[j] > N:
            continue
        all_wok.append(wok[i] + wok[j])

all_wok = sorted(all_wok)

visited = [-1] * (N + 1)
q = deque()

for w in all_wok:
    visited[w] = 1
    q.append(w)

while q:
    cur = q.popleft()
    for w in all_wok:
        if cur + w > N:
            break
        if visited[cur + w] == -1:
            q.append(cur + w)
            visited[cur + w] = visited[cur] + 1

print(visited[N])

File: 백준/Gold/13910. 개업/개업.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) throws IOException {

        st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());

        int[] wok = new int[M];
        List<Integer> allWok = new ArrayList<>();

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < M; i++) {
            int w = Integer.parseInt(st.nextToken());
            wok[i] = w;
            allWok.add(w);
        }

        for (int i = 0; i < M; i++) {
            for (int j = i + 1; j < M; j++) {
                if (wok[i] + wok[j] > N)
                    continue;
                allWok.add(wok[i] + wok[j]);
            }
        }

        Collections.sort(allWok);

        int[] visited = new int[N + 1];
        Queue<Integer> q = new ArrayDeque<>();

        Arrays.fill(visited, -1);

        for (int w: allWok) {
            visited[w] = 1;
            q.offer(w);
        }

        while (!q.isEmpty()) {
            int cur = q.poll();
            for (int w: allWok) {
                if (cur + w > N)
                    break;
                if (visited[cur + w] == -1) {
                    visited[cur + w] = visited[cur] + 1;
                    q.offer(cur + w);
                }
            }
        }

        System.out.println(visited[N]);

    } // end of main

} // end of class